File: challenges/2021/05-hydrothermalVenture/go/challenge.go
package challenge

import (
	"strconv"
	"strings"

	"github.com/codemicro/adventOfCode/lib/aocgo"
)

type Point struct {
	x int
	y int
}

func pointFromString(input string) (*Point, error) {
	sp := strings.Split(input, ",")
	x, err := strconv.Atoi(sp[0])
	if err != nil {
		return nil, err
	}
	y, err := strconv.Atoi(sp[1])
	if err != nil {
		return nil, err
	}
	return &Point{x: x, y: y}, nil
}

type Line [2]*Point

func parse(instr string) ([]Line, error) {
	var o []Line
	for _, line := range strings.Split(strings.TrimSpace(instr), "\n") {
		sp := strings.Split(line, " -> ")
		p1, err := pointFromString(sp[0])
		if err != nil {
			return nil, err
		}
		p2, err := pointFromString(sp[1])
		if err != nil {
			return nil, err
		}
		o = append(o, Line{p1, p2})
	}
	return o, nil
}

func removeDiagonalLines(lines []Line) []Line {
	var n int
	for _, line := range lines {
		if line[0].x == line[1].x || line[0].y == line[1].y {
			lines[n] = line
			n += 1
		}
	}
	return lines[:n]
}

func iteratePoints(line Line) chan *Point {

	p1 := line[0]
	p2 := line[1]

	delta_x := p2.x - p1.x
	delta_y := p2.y - p1.y

	var xStep, yStep int

	if delta_x > 0 {
		xStep = 1
	} else if delta_x < 0 {
		xStep = -1
	}

	if delta_y > 0 {
		yStep = 1
	} else if delta_y < 0 {
		yStep = -1
	}

	c := make(chan *Point)

	go func() {
		c <- p1
		lastPoint := *p1
		for lastPoint != *p2 {
			np := &Point{
				x: lastPoint.x + xStep,
				y: lastPoint.y + yStep,
			}
			c <- np
			lastPoint = *np
		}
		close(c)
	}()

	return c
}

func countOverlappingPoints(lines []Line) int {
	areas := make(map[Point]int)
	for _, line := range lines {
		for point := range iteratePoints(line) {
			t := *point
			areas[t] = areas[t] + 1
		}
	}

	var numberOverlaps int
	for _, n := range areas {
		if n > 1 {
			numberOverlaps += 1
		}
	}
	return numberOverlaps
}

type Challenge struct {
	aocgo.BaseChallenge
}

func (c Challenge) One(instr string) (interface{}, error) {
	lines, err := parse(instr)
	if err != nil {
		return nil, err
	}
	lines = removeDiagonalLines(lines)
	return countOverlappingPoints(lines), nil
}

func (c Challenge) Two(instr string) (interface{}, error) {
	lines, err := parse(instr)
	if err != nil {
		return nil, err
	}
	return countOverlappingPoints(lines), nil
}

File: challenges/2021/05-hydrothermalVenture/nim/challenge.nim
import std/tables
import std/sequtils
import std/strutils

type
    Point = object
        x: int
        y: int

    Line = (Point, Point)


proc pointFromString(instr: string): Point =
    let
        parts = instr.split(",")
        x = parts[0].parseInt
        y = parts[1].parseInt

    return Point(x: x, y: y)

proc parse(instr: string): seq[Line] =
    for line in instr.strip.splitLines:
        let parts = line.split(" -> ")
        result.add((
            pointFromString(parts[0]), pointFromString(parts[1])
        ))

proc removeDiagonalLines(lines: seq[Line]): seq[Line] =
    return lines.filter(
        proc(line: Line): bool = line[0].x == line[1].x or line[0].y == line[
                1].y,
    )

iterator iteratePoints(line: Line): Point =
    let
        (p1, p2) = line
        deltaX = p2.x - p1.x
        deltaY = p2.y - p1.y

    var
        xStep: int
        yStep: int

    if deltaX > 0:
        xStep = 1
    elif deltaX < 0:
        xStep = -1

    if deltaY > 0:
        yStep = 1
    elif deltaY < 0:
        yStep = -1

    var lastPoint = p1
    yield p1
    while lastPoint != p2:
        let np = Point(
            x: lastPoint.x + xStep,
            y: lastPoint.y + yStep,
        )
        yield np
        lastPoint = np

proc countOverlappingPoints(lines: seq[Line]): int =
    var areas = initTable[Point, int]()
    for line in lines:
        for point in iteratePoints(line):
            let nextValue = areas.getOrDefault(point, 0) + 1
            areas[point] = nextValue

    for count in areas.values:
        if count > 1:
            result += 1

proc partOne*(instr: string): int =
    return instr.
        parse.
        removeDiagonalLines.
        countOverlappingPoints

proc partTwo*(instr: string): int =
    return instr.
        parse.
        countOverlappingPoints
